src/ispc/kernels/lanczos3.ispc

```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
+    uniform WeightCollection * uniform horizontal_weight_collection = &cache->horizontal_weights;
+
+    // Accumulate only along the width for each pixel, sampling from the source image
+    // Results in the source image being downsampled to src_height X target_width
+    foreach_tiled(y =  0 ... src_height, x = 0 ... target_width) {
+        uint32 src_width_start = horizontal_weight_collection->starts[x];
+        uint32 num_horizontal_weights = horizontal_weight_collection->weight_counts[x];
+        float* horizontal_weights = horizontal_weight_collection->values[x];
+
+        float<3> color3 = {0.0f, 0.0f, 0.0f};
+        float<4> color4 = {0.0f, 0.0f, 0.0f, 0.0f};
+        for (uint32 i = 0; i < num_horizontal_weights; i++) {
+            float weight = horizontal_weights[i];
+            uint32 src_x = src_width_start + i;
+            uint64 src_read_address = (y * src_width + src_x) * num_channels;
+
+            if (num_channels == 3)
+                color3 += sample_3_channels(src_read_address, src_data) * weight;
+            else
+                color4 += sample_4_channels(src_read_address, src_data) * weight;
         }
+
+        uint64 scratch_write_address = (y * target_width + x) * num_channels;
+
+        if (num_channels == 3)
+            clean_and_write_3_channels(color3, scratch_write_address, scratch_space);
+        else
+            clean_and_write_4_channels(color4, scratch_write_address, scratch_space);
     }
-    col /= weight;
-    return col * 255;
-}
 
-export void resample(uniform uint32 width, uniform uint32 height, uniform uint32 stride, uniform uint8 num_channels, uniform uint32 target_width, uniform uint32 target_height, uniform const uint8 src_data[], uniform uint8 out_data[]) {
-    uniform Image src = {src_data, {width, height}};
-    uniform float<2> target_size = {(float)target_width, (float)target_height};
-    uniform float<2> inv_target_size = {1.0 / target_width, 1.0 / target_height};
+    // Accumulate the scratch space data along the height
+    // Downsamples the src_height X target_width image to target_height * target_width
+    foreach_tiled(y =  0 ... target_height, x = 0 ... target_width) {
+
+        uint32 src_height_start = vertical_weight_collection->starts[y];
+        uint32 num_vertical_weights = vertical_weight_collection->weight_counts[y];
+        float* vertical_weights = vertical_weight_collection->values[y];
 
-    foreach_tiled (y = 0 ... target_height, x = 0 ... target_width) {
-        float yf = y / target_size.y;
-        float xf = x / target_size.x;
+        float<3> color3 = {0.0f, 0.0f, 0.0f};
+        float<4> color4 = {0.0f, 0.0f, 0.0f, 0.0f};
+        for (uint32 i = 0; i < num_vertical_weights; i++) {
+            float weight = vertical_weights[i];
+            uint32 scratch_y = src_height_start + i;
+            uint64 scratch_read_address = (scratch_y * target_width + x) * num_channels;
```

Comment:
Are these gathers going to be hard to vectorize? Since this is the innermost loop, I wonder if the loop above could store the pixels transposed (so row-first) so that this increments by `num_channels` instead of by `num_channels * target_width` :)

(IOW: every scratch pixel is only stored once, but read many times)